password/src/Data/Password/Validate.hs

```diff
@@ -151,22 +163,32 @@ isValidPassword policy pre pass = null $ validatePassword policy pre pass
 -- []
 validatePassword :: PasswordPolicy -> CharSetPredicate -> Password -> [InvalidReason]
 validatePassword passwordPolicy@PasswordPolicy{..} charSetPredicate (Password password) =
-  catMaybes
-    [ isValidPolicy,
-      isTooShort,
-      isTooLong,
-      isUsingPolicyCharSetPredicate,
-      hasRequiredChar uppercaseChars Uppercase,
-      hasRequiredChar lowercaseChars Lowercase,
-      hasRequiredChar specialChars Special,
-      hasRequiredChar digitChars Digit
-    ]
+  -- There is no point in validating the password if either policy or predicate is invalid.
+  --
+  -- So we validate 'PasswordPolicy' and 'CharSetPredicate' first,
+  -- if they're valid, then validate 'Password'
+  firstNonEmpty
+   [ mconcat
+      [ validatePasswordPolicy passwordPolicy
+      , validateCharSetPredicate passwordPolicy charSetPredicate
+      ]
+   , catMaybes
+      [ isTooShort
+      , isTooLong
+      , isUsingValidCharacters
+      , hasRequiredChar uppercaseChars Uppercase
+      , hasRequiredChar lowercaseChars Lowercase
+      , hasRequiredChar specialChars Special
+      , hasRequiredChar digitChars Digit
+      ]
+   ]
```

Comment:
Split password validation in two steps. First, it' will validate the `PasswordPolicy` and `CharSetPredicate`. If either one is invalid, then it'll return `InvalidReason` withouth validating the `Passoword`. If both of them are valid, it'll then proceed to password validation.

This seemed natural to me since there's no point of returning password related `InvalidReason` if the policy is invalid.